src/meta/src/controller/catalog.rs

```diff
@@ -0,0 +1,389 @@
+            .sql_meta_store()
+            .expect("sql meta store is not initialized");
+        Ok(Self {
+            env,
+            inner: RwLock::new(CatalogControllerInner {
+                db: meta_store.conn,
+            }),
+        })
+    }
+}
+
+struct CatalogControllerInner {
+    db: DatabaseConnection,
+}
+
+impl CatalogController {
+    async fn notify_frontend(
+        &self,
+        operation: NotificationOperation,
+        info: NotificationInfo,
+    ) -> NotificationVersion {
+        self.env
+            .notification_manager()
+            .notify_frontend(operation, info)
+            .await
+    }
+}
+
+impl CatalogController {
+    pub fn snapshot(&self) -> MetaResult<()> {
+        todo!("snapshot")
+    }
+
+    async fn create_object(
+        txn: &DatabaseTransaction,
+        obj_type: ObjectType,
+        owner_id: UserId,
+    ) -> MetaResult<object::Model> {
+        let active_db = object::ActiveModel {
+            oid: Default::default(),
+            obj_type: ActiveValue::Set(obj_type),
+            owner_id: ActiveValue::Set(owner_id as _),
+            initialized_at: Default::default(),
+            created_at: Default::default(),
+        };
+        Ok(active_db.insert(txn).await?)
+    }
+
+    pub async fn create_database(&self, db: PbDatabase) -> MetaResult<NotificationVersion> {
+        let inner = self.inner.write().await;
+        let txn = inner.db.begin().await?;
+        let owner_id = db.owner;
+
+        let db_obj = Self::create_object(&txn, ObjectType::Database, owner_id).await?;
+        let mut db: database::ActiveModel = db.into();
+        db.database_id = ActiveValue::Set(db_obj.oid);
+        let db = db.insert(&txn).await?;
+
+        let mut schemas = vec![];
+        for schema_name in iter::once(DEFAULT_SCHEMA_NAME).chain(SYSTEM_SCHEMAS) {
+            let schema_obj = Self::create_object(&txn, ObjectType::Schema, owner_id).await?;
+            let schema = schema::ActiveModel {
+                schema_id: ActiveValue::Set(schema_obj.oid),
+                name: ActiveValue::Set(schema_name.into()),
+                database_id: ActiveValue::Set(db.database_id),
+            };
+            let schema = schema.insert(&txn).await?;
+            schemas.push(ObjectModel(schema, schema_obj).into());
+        }
+        txn.commit().await?;
+
+        let mut version = self
+            .notify_frontend(
+                NotificationOperation::Add,
+                NotificationInfo::Database(ObjectModel(db, db_obj).into()),
+            )
+            .await;
+        for schema in schemas {
+            version = self
+                .notify_frontend(NotificationOperation::Add, NotificationInfo::Schema(schema))
+                .await;
+        }
+
+        Ok(version)
+    }
+
+    /// This function will list all the objects that are using the given one. It runs a recursive CTE to find all the dependencies.
+    /// The cte and the query is as follows:
+    /// ```sql
```

Comment:
BTW, can we make this a doc test?

https://www.sea-ql.org/sea-orm-tutorial/ch01-08-sql-with-sea-query.html#testing-and-debugging

Reply:
Yes sure, let me add one for it.